### ISSUE TITLE
Latest version delve does not properly close or restart. 0.6.84 Works fine.

### ISSUE THREAD

**Issue by antihax** [Microsoft/vscode-go#1814]
Having issues on Ubuntu with Go extension 0.6.85 and delve debugger on multiple projects. Looks related to #438. This started after upgrading to 0.6.85 and works fine when downgrading to 0.6.84.

**Issues:**
1. Have to hit stop twice to stop debugging, on occasion the process keeps running anyway but vscode says it is not.
2. Hitting restart does nothing, process keeps running and logging to the window.
3. If the code will not build, resolving the build then trying to launch the debugger results in and error: `There is already a debug configuration "Launch" running.` when there is nothing running. Have to restart vscode to fix.

launch.json
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Launch",
            "type": "go",
            "request": "launch",
            "mode": "debug",
            "remotePath": "",
            "port": 2345,
            "host": "127.0.0.1",
            "cwd": "${workspaceRoot}",
            "program": "${workspaceRoot}/cmd/main.go",
            "args": [],
        },
    ]
}
```

debug.log (DisconnectRequest appears after first stop request, nothing after second)
```
11:52:55 AM, 7/29/2018
InitializeRequest
InitializeResponse
Using GOPATH: /home/antihax/go
Running: /home/antihax/go/bin/dlv debug github.com/antihax/eve-marketwatch/cmd --headless=true --listen=127.0.0.1:2345 --log=true --wd=/home/antihax/go/src/github.com/antihax/eve-marketwatch --
API server listening at: 127.0.0.1:2345
time="2018-07-29T11:52:56-07:00" level=info msg="launching process with args: [/home/antihax/go/src/github.com/antihax/eve-marketwatch/cmd/debug]" layer=debugger
InitializeEvent
ConfigurationDoneRequest
ContinueRequest
ContinueResponse
ThreadsRequest
time="2018-07-29T11:52:56-07:00" level=debug msg=continuing layer=debugger
DisconnectRequest
```
Both delve and debugged process are still running.

**Comment by cpwc** [comment]
Same issue here on macOS with `0.6.85` had to downgrade to `0.6.84` for debugger work correctly.

**Comment by mvrhov** [comment]
There is probably more of them like that:
* if the build fails you cannot start again. you get that the Debug process is already running error.
* the restart is now sometimes so slow, that process restart manages to launch multiple instances.
* you have to press stop button multiple times for it to register the stop or to stop debug process I dno't know which one is true

**Comment by ramya-rao-a** [comment]
Can someone here share a small sample code where I can try and get a repro of this?

**Comment by antihax** [comment]
This appears to cause some of the issues mentioned above.
```
package main

import (
	"sync"
	"time"
)

func main() {
	wg := sync.WaitGroup{}
	wg.Add(1)
	go func() {
		for {
			time.Sleep(time.Second)
		}
	}()
	wg.Wait()
}
```

**Comment by gencer** [comment]
I'm on windows and this happens.

**Comment by ramya-rao-a** [comment]
@antihax I tried your code on Windows. On ending the debug session using the stop button, debugging stops as expected.

Whats changed between 0.6.84 and 0.6.85 is the below:
- In 0.6.84 when you stop the debugging session, we forcefully killed the debug process that was spawned for delve and all its children.
- In 0.6.85 when you stop the debugging session, instead of the force kill, we send a [detach](https://godoc.org/github.com/derekparker/delve/service/rpc1#RPCServer.Detach) request to delve. Delve then kills all processes. Once delve gets back to us saying that detach is complete, we (the Go extension) then request VS Code to disconnect which is when the debug tool bar goes away

From the descriptions in this issue, it looks like the detach by delve is not getting completed. I have added a few more logs to the code to confirm this. Can folks here try the below?
- Download https://github.com/Microsoft/vscode-go/blob/master/Go-latest.vsix
- Run `code --install-extension Go-latest.vsix` 
- If the above fails with `Error: end of central directory record signature not found`, then clone this repo and use the `Go-latest.vsix` file from the cloned repo
- Reload VS Code
- Add `"trace": "verbose"` to your debug configuration in the launch.json file.
- Try to repro the issue

Look for "DetachRequest", "DetachResponse" and "DisconnectRequest to parent" in the logs and share what you find here.

**Comment by antihax** [comment]
With above code and Go-latest.vsix from cloned repo.
First stop press `DisconnectRequest` appears below. Second press, nothing additional.

```
Verbose logs are written to:
/tmp/vscode-go-debug.txt
2:46:03 PM, 8/1/2018
InitializeRequest
InitializeResponse
Using GOPATH: /home/antihax/go
Running: /home/antihax/go/bin/dlv debug github.com/antihax/test --headless=true --listen=127.0.0.1:2345 --log=true --
API server listening at: 127.0.0.1:2345
time="2018-08-01T14:46:03-07:00" level=info msg="launching process with args: [/home/antihax/go/src/github.com/antihax/test/debug]" layer=debugger
InitializeEvent
ConfigurationDoneRequest
ContinueRequest
ContinueResponse
ThreadsRequest
time="2018-08-01T14:46:04-07:00" level=debug msg=continuing layer=debugger
DisconnectRequest
```

**Comment by ramya-rao-a** [comment]
@antihax We also send a Halt request before trying to Detach for which I forgot to add logs earlier. Can you try the instructions at https://github.com/Microsoft/vscode-go/issues/1814#issuecomment-409729999 again?

**Comment by antihax** [comment]
On first stop. 
```
DisconnectRequest
HaltRequest
```

On second, no change to log but the controls go away. Both debug and dlv are still running.

```
 98102 tty2     Sl+    0:00 /home/antihax/go/bin/dlv debug github.com/antihax/test --headless=true --listen=127.0.0.1:2345 --log=true --
 98188 tty2     Sl     0:00 /home/antihax/go/src/github.com/antihax/test/debug
```

**Comment by ramya-rao-a** [comment]
This shows that the halt request is not getting completed by delve. I still cant repro this, so not sure what the next steps would be

@aarzilli any thoughts on why the Halt request made to delve would not complete and call the given callback?

**Comment by aarzilli** [comment]
I'd like to see the output of `--log --log-output=rpc`.

**Comment by rain2fog** [comment]
same problem with me 
environment: ubuntu 16.04 /  vscode 1.25.1 / vscode-go 0.6.85
I have to downgrade to 0.6.84, then it worked.

**Comment by ramya-rao-a** [comment]
@antihax I've updated the `Go-latest.vsix` to have the ability to configure the `--log-output`. Can you re-do the steps from https://github.com/Microsoft/vscode-go/issues/1814#issuecomment-409729999, add `"logOutput": "rpc"` to your debug configuration in launch.json file and the share the logs so that @aarzilli can take a look?

**Comment by ramya-rao-a** [comment]
I am able to get logs after the `HaltRequest`

<img width="1059" alt="screen shot 2018-08-02 at 4 54 45 pm" src="https://user-images.githubusercontent.com/16890566/43617087-d0d77a56-9674-11e8-92c3-e5800dbd4892.png">

@aarzilli I log the text `HaltRequest` just before sending the halt command to delve, as you can see in my screenshot above. In @antihax's case, delve doesn't return any logs after the attempt to halt. In my case, you can see the logs form delve after the attempt to halt

@antihax Do you have the latest dlv? Can you update it just to be sure? You can run `Go: Install/Update Tools` and select `dlv` to update it.

**Comment by antihax** [comment]
```
Delve Debugger
Version: 1.0.0
Build: $Id: c98a142125d0b17bb11ec0513bde346229b5f533 $
```
Same thing after running the update above.

**Comment by ramya-rao-a** [comment]
@antihax And you are trying this with the sample code from https://github.com/Microsoft/vscode-go/issues/1814#issuecomment-409424205 correct?

**Comment by antihax** [comment]
Yes.

**Comment by antihax** [comment]
I am using the code above. 

`go version go1.10.3 linux/amd64` on `Ubuntu 18.04.1 LTS`

VS Code is:
```
Version: 1.25.1
Commit: 1dfc5e557209371715f655691b1235b6b26a06be
Date: 2018-07-11T15:40:20.190Z
Electron: 1.7.12
Chrome: 58.0.3029.110
Node.js: 7.9.0
V8: 5.8.283.38
Architecture: x64
```

**Comment by ramya-rao-a** [comment]
@antihax I set up a ubuntu environment via parallels on my Mac and tried your configration, still no repro. But then I realized I was using the Insiders build of VS Code. Then I tried the stable build, and there you go I see the same logs as you. Nothing after `HaltRequest`!!

Can you try the latest [Insiders](https://code.visualstudio.com/insiders) and let me know if you still see the issue?

**Comment by aarzilli** [comment]
In @antihax logs I see that VSCode is sending a ListGoroutines request right after the continue command, that request is synchronous and can't be executed until continue terminates and will also block any other command from going through (including halt commands).

**Comment by Tusharsb** [comment]
@ramya-rao-a 
> Can you try adding a breakpoint and then see if you are able to stop the debugging session properly?

I am able to reproduce the issue even with breakpoints ON. I am using a different code as follows
```
package main

import (
	"fmt"
	"log"
	"net/http"

	"github.com/gorilla/mux"
)

// our main function
func main() {
	router := mux.NewRouter()
	router.HandleFunc("/", IncomingMessage).Methods("POST")
	log.Fatal(http.ListenAndServe(":8002", router))
}

func IncomingMessage(w http.ResponseWriter, r *http.Request) {
	fmt.Println("hello world")
}
```
with breakpoint set on second last line

`go version go1.10.3 linux/amd64`

**Comment by imkos** [comment]
reproduce the issue:  
1. to start debugging on  no "main"  package, This will certainly fail to start.
2. After to start debugging  "main"  package,  There is already a debug configuration "Launch" running.

**Comment by moitias** [comment]
Seeing this on Windows too, and it seems to be related to how build failures are handled at least on my configuration. The basic steps to reproduce are:

0) (optional) Run code succesfully, debug succesfully
1) Try to run unbuildable code
2) Expect to be able to run/debug again.

Tested on the latest release Code and the go-latest.vsix (commit b4c4698) from the repo - doesn't work - and with 0.6.84 linked above - works fine. 

Logs from both a succesful run and the logs after running unbuildable code from both versions below. No RPC logging on the 0.6.84 as the launch configuration flag doesn't work in the older plugin.

Code to run:
```
package main

import (
	"fmt"
)

func main() {
	fmt.Println("HELLO")
	// uncomment this line and f5 to break debugging
}
```

Runs and debugs fine, repeatedly. Logs on a succesful session are as follows:
```
Verbose logs are written to:
C:\Users\matias\AppData\Local\Temp\vscode-go-debug.txt
19:40:30, 2018-8-4
InitializeRequest
InitializeResponse
Using GOPATH: c:\users\matias\code\worldo
Running: c:\users\matias\code\worldo\bin\dlv.exe debug cmd\test --headless=true --listen=127.0.0.1:2345 --log=true --log-output=rpc --
time="2018-08-04T19:40:31+03:00" level=info msg="Using API v1" layer=rpc
API server listening at: 127.0.0.1:2345
InitializeEvent
ConfigurationDoneRequest
ContinueRequest
ContinueResponse
ThreadsRequest
time="2018-08-04T19:40:31+03:00" level=debug msg="<- RPCServer.GetVersion(api.GetVersionIn{})" layer=rpc
time="2018-08-04T19:40:31+03:00" level=debug msg="-> *api.GetVersionOut{\"DelveVersion\":\"Version: 1.0.0\\nBuild: $Id: c98a142125d0b17bb11ec0513bde346229b5f533 $\",\"APIVersion\":1} error: \"\"" layer=rpc
time="2018-08-04T19:40:31+03:00" level=debug msg="(async 2) <- RPCServer.Command(*api.DebuggerCommand{\"name\":\"continue\",\"ReturnInfoLoadConfig\":null})" layer=rpc
time="2018-08-04T19:40:31+03:00" level=debug msg="<- RPCServer.ListGoroutines(<nil>null)" layer=rpc
time="2018-08-04T19:40:31+03:00" level=debug msg="-> *[]*api.Goroutine[] error: \"\"" layer=rpc
HELLO
time="2018-08-04T19:40:31+03:00" level=debug msg="(async 2) -> *api.DebuggerState{\"Running\":false,\"Threads\":null,\"NextInProgress\":false,\"exited\":true,\"exitStatus\":0,\"When\":\"\"} error: \"\"" layer=rpc
continue state {"Running":false,"Threads":null,"NextInProgress":false,"exited":true,"exitStatus":0,"When":""}
TerminatedEvent
DisconnectRequest
HaltRequest
time="2018-08-04T19:40:31+03:00" level=debug msg="(async 4) <- RPCServer.Command(*api.DebuggerCommand{\"name\":\"halt\",\"ReturnInfoLoadConfig\":null})" layer=rpc
time="2018-08-04T19:40:31+03:00" level=debug msg="(async 4) -> *api.DebuggerStatenull error: \"Process 27040 has exited with status 0\"" layer=rpc
HaltResponse
DetachRequest
time="2018-08-04T19:40:31+03:00" level=debug msg="<- RPCServer.Detach(booltrue)" layer=rpc
DetachResponse
DisconnectRequest to parent

```
The same (build success) on 0.6.84:
```
Verbose logs are written to:
C:\Users\matias\AppData\Local\Temp\vscode-go-debug.txt
19:49:57, 2018-8-4
InitializeRequest
InitializeResponse
Using GOPATH: c:\users\matias\code\worldo
Running: c:\users\matias\code\worldo\bin\dlv.exe debug cmd\test --headless=true --listen=127.0.0.1:2345 --log=true --
API server listening at: 127.0.0.1:2345
time="2018-08-04T19:49:57+03:00" level=info msg="launching process with args: [c:\\Users\\matias\\code\\worldo\\src\\cmd\\test\\debug]" layer=debugger
InitializeEvent
ConfigurationDoneRequest
ContinueRequest
ContinueResponse
ThreadsRequest
time="2018-08-04T19:49:58+03:00" level=debug msg=continuing layer=debugger
HELLO
continue state {"Running":false,"Threads":null,"NextInProgress":false,"exited":true,"exitStatus":0,"When":""}
TerminatedEvent
DisconnectRequest

```



After uncommenting the commented line - causing an error ('unexpected to at end of statement') while building - it is impossible to re-run the debugger. 

F5 seems to be completely ignored, while running from the Debug pane results in an 'There is already a debug configuration running' error message. Stop debugging is active, but does not result in anything while restarting results in the same error. Trying to terminate/stop/anything from command palette does not help either - the only workarounds seem to be a complete restart after each build failure (which is pretty much unusable) or downgrading to 0.6.84.

Logs from build failure are as follows:

```
Verbose logs are written to:
C:\Users\matias\AppData\Local\Temp\vscode-go-debug.txt
19:41:01, 2018-8-4
InitializeRequest
InitializeResponse
Using GOPATH: c:\users\matias\code\worldo
Running: c:\users\matias\code\worldo\bin\dlv.exe debug cmd\test --headless=true --listen=127.0.0.1:2345 --log=true --log-output=rpc --
# cmd/test
.\main.go:9:12: syntax error: unexpected to at end of statement
exit status 2
Process exiting with code: 1
Delve is closed
DisconnectRequest
HaltRequest
``` 

The same (build failure) on 0.6.84: 
```
Verbose logs are written to:
C:\Users\matias\AppData\Local\Temp\vscode-go-debug.txt
19:50:40, 2018-8-4
InitializeRequest
InitializeResponse
Using GOPATH: c:\users\matias\code\worldo
Running: c:\users\matias\code\worldo\bin\dlv.exe debug cmd\test --headless=true --listen=127.0.0.1:2345 --log=true --
# cmd/test
.\main.go:9:12: syntax error: unexpected this at end of statement
exit status 2
Process exiting with code: 1
Delve is closed
DisconnectRequest
```

**Comment by ramya-rao-a** [comment]
> In @antihax logs I see that VSCode is sending a ListGoroutines request right after the continue command, that request is synchronous and can't be executed until continue terminates and will also block any other command from going through (including halt commands).

@aarzilli That explains why the halt command never makes it through. Any reason why continue wouldn't terminate? 

@moitias Thanks for the detailed steps. The `There is already a debug configuration running` is related to the previous session not completing properly which is due to what @aarzilli mentioned above

**Comment by aarzilli** [comment]
> That explains why the halt command never makes it through. Any reason why continue wouldn't terminate?

The program in question is an infinite loop, if the halt doesn't make it through it will never terminate on its own.

**Comment by ramya-rao-a** [comment]
> if the halt doesn't make it through it will never terminate on its own.

@aarzilli You mean the program will not terminate on its own or the `continue` command? What is the suggested way to go about this for delve clients?

**Comment by gregmarr** [comment]
I am on Mac OS and am also seeing all of the issues reported here: having to click twice to stop, processes sticking around that I have to kill on my own using Activity Monitor, and the error about debugging already running for which I have to restart VSCode.  After I started seeing this I updated Delve and the Go tools, so they're current.

**Comment by ramya-rao-a** [comment]
I'll look into that, thanks.

**Comment by vasmik** [comment]
Mac OS - same issue. Downgraded to 0.6.84 :(

**Comment by ramya-rao-a** [comment]
> Not sending requests while the target is running.

This doesn't seem trivial. There is no way way for the debug adapter to know if the target is still running.

Instead, I have brought back the code to kill the process manually, but only if we don't hear back from the halt request made to delve. @moitias This will fix your issue as well.

Everyone here, please try the below
- Download https://github.com/Microsoft/vscode-go/blob/master/Go-latest.vsix
- Run `code --install-extension Go-latest.vsix` 
- If the above fails with `Error: end of central directory record signature not found`, then clone this repo and use the `Go-latest.vsix` file from the cloned repo
- Reload VS Code
- Try and debug your code now

**Comment by aarzilli** [comment]
> This doesn't seem trivial. There is no way way for the debug adapter to know if the target is still running.

The target is running from the time you make a Command request to the time you receive a response for it.

**Comment by ramya-rao-a** [comment]
Thanks @aarzilli. I've included a check to not make the `ListRoutines` call if a previous continue request is pending. Regardless I'll keep the code to kill manually if halt doesnt return in time anyway. That should act as a catch all on other unforeseen issues.

All,
If you havent already, please update to the latest beta version by following the steps in https://github.com/Microsoft/vscode-go/issues/1814#issuecomment-411566434 and let me know if the issue has resolved for you.

**Comment by cpwc** [comment]
@ramya-rao-a the latest beta version works perfectly.

**Comment by antihax** [comment]
I am still seeing the same issue hanging on `ListGoroutines` with vscode-go v0.6.68-beta.3
```...
InitializeEvent
ConfigurationDoneRequest
ContinueRequest
ContinueResponse
ThreadsRequest
time="2018-08-11T07:13:37-07:00" level=debug msg="<- RPCServer.GetVersion(api.GetVersionIn{})" layer=rpc
time="2018-08-11T07:13:37-07:00" level=debug msg="-> *api.GetVersionOut{\"DelveVersion\":\"Version: 1.0.0\\nBuild: $Id: c98a142125d0b17bb11ec0513bde346229b5f533 $\",\"APIVersion\":1} error: \"\"" layer=rpc
time="2018-08-11T07:13:37-07:00" level=debug msg="(async 2) <- RPCServer.Command(*api.DebuggerCommand{\"name\":\"continue\",\"ReturnInfoLoadConfig\":null})" layer=rpc
time="2018-08-11T07:13:37-07:00" level=debug msg="<- RPCServer.ListGoroutines(<nil>null)" layer=rpc
DisconnectRequest
HaltRequest```

**Comment by Evasion** [comment]
Appears to still be an issue in 0.6.88

**Comment by ramya-rao-a** [comment]
@Evasion Please add `"trace": "verbose"` to your debug configuration, share the output logs and a sample code that I can use to repro

**Comment by ramya-rao-a** [comment]
@Evasion, if you do see it again, please log a new issue. This thread is already long :)
